websocket_client.py

```diff
@@ -0,0 +1,463 @@
+import argparse
+import asyncio
+import datetime
+import json
+import logging
+import os
+import signal
+import sys
+import urllib.parse
+from typing import Any, AsyncGenerator, Awaitable, Literal
+
+import aiohttp
+import numpy as np
+import pyee.asyncio
+import sounddevice
+from websockets import exceptions as ws_exceptions
+from websockets.asyncio import client as ws_client
+
+
+class LocalAudioSink:
+    """AudioSink that plays to the default audio device."""
+
+    def __init__(self, sample_rate: int = 48000, num_channels: int = 1) -> None:
+        self._sample_rate = sample_rate
+        self._num_channels = num_channels
+        self._buffer: bytearray = bytearray()
+
+        def callback(outdata: np.ndarray, frame_count, time, status):
+            output_frame_size = len(outdata) * 2
+            next_frame = self._buffer[:output_frame_size]
+            self._buffer[:] = self._buffer[output_frame_size:]
+            if len(next_frame) < output_frame_size:
+                next_frame += b"\x00" * (output_frame_size - len(next_frame))
+            outdata[:] = np.frombuffer(next_frame, dtype="int16").reshape(
+                (frame_count, num_channels)
+            )
+
+        self._stream = sounddevice.OutputStream(
+            samplerate=sample_rate,
+            channels=num_channels,
+            callback=callback,
+            device=None,
+            dtype="int16",
+            blocksize=sample_rate // 100,
+        )
+        self._stream.start()
+        if not self._stream.active:
+            raise RuntimeError("Failed to start streaming output audio")
+
+    def write(self, chunk: bytes) -> None:
```

Comment:
"""
        Write audio data to the playback buffer.

        Args:
            chunk (bytes): Raw audio data to be played
        """